### ISSUE TITLE
[New Check] Warn when builtins might mask return codes

### ISSUE THREAD

**Issue by sboyd-m** [koalaman/shellcheck#2439]
#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/sh

foo()
{
    set -x
    false
}

bar()
{
    set -x
    false
    set +x
}

if foo; then echo success; else echo fail; fi
if bar; then echo success; else echo fail; fi

```

#### Here's what shellcheck currently says:
Nothing

#### Here's what I wanted or expected to see:
Warning: Shell builtins mask return codes

With any normal command, I would consider this expected behavior. However, since `set` isn't really a "normal" command, it is unexpected that it could mask return codes and break the conditional. It's kind of a corner case, but I just ran into it in the field :)
There may also be other areas where this could happen outside my sample here, so I'm not sure about the scope of the issue.

**Comment by SuperSandro2000** [comment]
It is not masking the return code. The last command is just successful and you are missing `set -eou pipefail`.

**Comment by sboyd-m** [comment]
You are clearly not understanding the issue here. Please read the entire post before commenting.

**Comment by SuperSandro2000** [comment]
You think that bash handles `set` any different to any other command which is not the case. It might be technically a builtin but as far as I know that has no influence on return codes. It would be unexpected to handle it any different to any other command in a function.

**Comment by sboyd-m** [comment]
I do not "think" that bash handles set any differently. Again, you did not read the issue.

**Comment by lukebakken** [comment]
Why should `shellcheck` flag a script that is doing what you tell it to do? How could this case be detected generally? `set +x` is a perfectly valid thing to do.

Finally, you're being unhelpful by saying "you did not read the issue". If you think that the issue is being misunderstood, take the time to explain why.

**Comment by sboyd-m** [comment]
>Why should shellcheck flag a script that is doing what you tell it to do?

This kind of question is so rhetorical as to be completely unconstructive. There are many "gotchas" and footguns in many languages, especially so in shell scripts. There are already many, many warnings in shellcheck that are of the form "I see what you are _trying_ to do here, but you actually are telling the shell to do _X_, so here is a better way, or here is the 'correct' way". I am asking for this check to be added, as one of these gotcha warnings.

>How could this case be detected generally?

In exactly the way I described in the OP. If someone is checking a return code _right after using `set`_, they are 99.9999% trying to check the return code of the command right before `set`, and not whether `set` succeeded.

A more trivial example might be:
```
set -x
a
long
series
of
commands
set +x
if [ "$?" = "0" ]...
```

I think anyone seeing that might find it obvious that whoever wrote it made a mistake. Of course, this more trivial example would already sort of be caught be shellcheck, since it warns about checking "$?" in an if statement as opposed to just checking the command directly.

This is why I posted the more complex version, which is a simplified version of what I ran into in the field when debugging an archaic script that was misbehaving. I usually turn to shellcheck to look at very old scripts, since it is very good at finding little gotchas that probably are creating the bug I'm investigating.

Luckily, since shellcheck is fairly advanced, I feel like it might be able to see a bit deeper into how scripts run to detect this more complex case. Here is the broad logic:
Say, if the last command in a function/alias being checked by the user is a shell builtin, then assume it is a mistake and offer a warning. The broad logic is very straightforward, but of course things usually wind up being more difficult than they appear, as I said in OP.

**Comment by TinCanTech** [comment]
> if [ "$?" = "0" ]...

`shellcheck` won't like that either ..

**Comment by lukebakken** [comment]
Thanks for the detailed explanation. @SuperSandro2000 will appreciate it as well. I'm glad I didn't have to debug a failing script due to `set +x` being the last statement in a function. This is a useful cautionary tale. It might also be helpful to remind people watching this issue that `set -x` / `set +x` enables and disables tracing of commands and thus the return value is next-to meaningless (like you point out).

_Now_ I can see how detecting this case would be useful. I'm not sure if it should be extended to [all POSIX sh builtins](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_14). I will say I've never (knowingly) checked the return value of any.

**Comment by sboyd-m** [comment]
Indeed, it can get quite hairy. I would think some builtins are fair game, `exec` comes to mind, since if it returns _at all_ theres a problem, so it would be fair to check that one in a script. Also the most obvious one being `return`, which _really should_ be checked.

This is why I mentioned concerns about scope at the end of OP, because it could easily turn into having some semi-arbitrary subset of shell builtins being checked, or different builtin lists for different shells (many shells have `true` and `false` as builtins for example).

To me though, `set` is one command that really stands out as never being worth checking, as I'm sure the buggy script author also felt.

**Comment by e-kwsm** [comment]
All you need to do is run `set -e` beforehand.

**Comment by sboyd-m** [comment]
`set -e` would not address this issue at all. Are you just trying to close old issues?

**Comment by e-kwsm** [comment]
It is not only `set` that discards return code of the previous command:

```sh
f() {
  …
  echo 'function f finished'  # this builtin masks return code!
}

g() {
  pushd foo || return
  …
  popd  # ditto
}
```

and you have to `set -e` (#409), or
```sh
this-command-may-fail || return
```

**Comment by wileyhy** [comment]
sboyd-m:

  "You are clearly not understanding the issue here."

Really? Come on.

  "If someone is checking a return code right after using set, they are
99.9999% trying to check the return code of the command right before
set...."

That's an thoroughly unprovable assumption. And it's a false assertion. The
fact is `set +x` is valid code, meaning, it is not in and of itself invalid
code.

Why try to be clairvoyant when it's a science and highly specific and fully
discoverable answers are a thing: why guess?

People shortcut conditional tests with always-true statements all the time.
Have you read the shell scripts on your Linux box? (*cough* ALSA! *cough*)
It's not great technique but if you have no time to fix something else....
It happens.

  "Say, if the last command in a function/alias being checked by the user
is a shell builtin, then assume it is a mistake and offer a warning."

What if the last command of a function was `return`?   o_O   Would that be
a mistake? Of course not.

What if the last command of a function was `exit` or `kill`? Should all
`trap` functions be flagged with a warning? Why should idiomatic shell code
get flagged as wrong? The exit code of `exit` is something people look for.

Why do people write built-ins anyway? Because they're faster.

Do shells still write new built-ins? Yes.

Can people write their own built-ins? Yes.

Do companies write and use their own built-ins? They certainly can. It's a
fair bet that out there in the wild somewhere a few companies have been
doing just that since the feature was released. It's possible the feature
exists as a result of a feature request.

In Bash, built-ins are considered "commands." In which shells, if any, are
built-ins expressly categorized as something other than "commands?"

Authorial intent is notoriously difficult to glean correctly.

Why are there certain shellcheck rules that appear to say, 'you were
probably intending to do blah - here's how to do that?' Because people ask
the same 100 questions on IRC constantly, and typing the same answer 50
times becomes tedious.

Is "should a built-in never precede a conditional" one of those
oft-repeated questions? The last time I checked, no.

Have I ever heard that question before? Not that I recall.

I'm not sure it's worth continuing this discussion.

Wiley
